docs/EpochBasedMemoryManagement.md

```diff
@@ -94,35 +107,94 @@ typedef struct _ebpf_epoch_allocation_header
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_EXIT_EPOCH, /// Remove the ebpf_epoch_state_t from the per-CPU thread list.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_RUNDOWN_IN_PROGRESS, /// Rundown is in progress. Stop processing messages.
+    EBPF_EPOCH_CPU_MESSAGE_TYPE_IS_FREE_LIST_EMPTY, /// Is the free list empty?
+} ebpf_epoch_cpu_message_type_t;
+
+/**
+ * @brief Message sent to a CPU to perform an action.
+ * Message can be one of the following:
+ * 1. Compute the minimum epoch of all threads on the CPU and send the message to the next CPU.
+ * If this is the last CPU, then set the _ebpf_release_epoch to the minimum epoch and
+ * send a message to the first CPU to rearm the timer.
+ * 2. Remove all entries from the per-CPU free list that have an epoch that is before released_epoch, rearm
+ * the timer if needed, and send the message to the next CPU. If this is the last CPU, then stop.
+ * 3. Call ebpf_epoch_exit() on behalf of the other CPU.
+ * 4. Set the _ebpf_epoch_cpu_entry_t.rundown_in_progress flag to true and send the message to the next CPU.
+ * If this is the last CPU, then signal the caller that rundown is complete.
+ */
+typedef struct _ebpf_epoch_cpu_message
+{
+    ebpf_epoch_cpu_message_type_t message_type;
+    union
+    {
+        struct
+        {
+            uint64_t epoch;
+        } compute_epoch;
+        struct
+        {
+            uint8_t unused;
+        } rearm_timer;
+        struct
+        {
+            ebpf_epoch_state_t* epoch_state;
+        } exit_epoch;
+        struct
+        {
+            uint8_t unused;
+        } rundown_in_progress;
+        struct
+        {
+            bool is_empty;
+        } is_free_list_empty;
+    } message;
+    KEVENT completion_event;
+} ebpf_epoch_cpu_message_t;
+```
 
-## Exceptional cases
+### ebpf_epoch_cpu_message_type_t
+This message contains the oldest known epoch. It is sent to each CPU, which then
+queries the epoch for each thread queued on that CPU and sets the epoch in the
+message to the minimum of the local minima and the minima in the message. The
+message is then forwarded to the next CPU. The last CPU then sets
+_ebpf_release_epoch to the minima minus 1 and sends a re-arm timer to CPU 0.
 
-There are a few exceptional cases handled in the epoch module.
+### EBPF_EPOCH_CPU_MESSAGE_TYPE_REARM_TIMER
+This message instructs the CPU to clear the timer-armed flag, releases any items
+in the free list that are eligible for reclamation, re-arms the timer if
+need, and forwards the message to the next CPU. The last CPU then sends a epoch
+complete message to CPU 0.
 
-### Stale free lists
+### EBPF_EPOCH_CPU_MESSAGE_TYPE_COMPUTE_EPOCH_COMPLETE
+This message is sent to CPU 0 to indicate that computation of the
+_ebpf_release_epoch is complete and that future epoch computation can now be
+initiated. In addition, it signals the KEVENT associated with the message to
+signal any waiting threads that the operation is completed.
```

Comment:
In my view, this internal documentation would be better in the doxygen description on the actual enum value.
That puts it in context in the actual file where the code is.